cachi2/core/package_managers/rubygems/parser.py

```diff
@@ -0,0 +1,131 @@
+import logging
+import re
+from typing import Iterable, Optional, Union
+
+import pydantic
+from gemlock_parser.gemfile_lock import GEM, GIT, PATH, Gem, GemfileLockParser  # type: ignore
+
+from cachi2.core.errors import PackageRejected, UnexpectedFormat
+from cachi2.core.rooted_path import PathOutsideRoot, RootedPath
+
+log = logging.getLogger(__name__)
+
+GEMFILE_LOCK = "Gemfile.lock"
+
+
+class GemMetadata(pydantic.BaseModel, extra="forbid"):
+    """
+    Base class for gem metadata.
+
+    Attributes:
+        name:       The name of the gem.
+        version:    The version of the gem.
+    """
+
+    name: str
+    version: str
+
+
+class GemDependency(GemMetadata):
+    """
+    Represents a gem dependency.
+
+    Attributes:
+        source:     The source URL of the gem.
+        checksum:   The checksum of the gem.
+    """
+
+    source: str
+    checksum: Optional[str] = None
+
+
+class GitDependency(GemMetadata):
+    """
+    Represents a git dependency.
+
+    Attributes:
+        url:        The URL of the git repository.
+        revision:   Commit hash.
+        branch:     The branch of the git repository.
+    """
+
+    url: str
+    revision: str
+    branch: Optional[str] = None
+
+
+class PathDependency(GemMetadata):
+    """
+    Represents a path dependency.
+
+    Attributes:
+        path:       Subpath from package root.
+    """
+
+    path: str
+
+
+BundlerDependency = Union[GemDependency, GitDependency, PathDependency]
+
+
+def parse_gemlock(package_dir: RootedPath) -> list[BundlerDependency]:
+    """Parse a Gemfile.lock file and return a list of dependencies."""
+    gemlock_path = package_dir.join_within_root(GEMFILE_LOCK)
+    if not gemlock_path.path.exists():
+        reason = f"Gemfile.lock not found at {gemlock_path.path}"
+        solution = "Please ensure the Gemfile.lock file is present in the package directory"
+        raise PackageRejected(reason=reason, solution=solution)
+
+    parser = GemfileLockParser(gemlock_path.path)
+    parsed_gems: Iterable[Gem] = parser.all_gems.values()
+    log.info("Bundler version - %s", parser.bundled_with)
+
+    dependencies: list[BundlerDependency] = []
+    for gem in parsed_gems:
+        if gem.type == GEM:
+            gem_dep = GemDependency(
+                name=gem.name,
+                version=gem.version,
+                source=gem.remote,
+                checksum=None,
+            )
+            dependencies.append(gem_dep)
+        elif gem.type == GIT:
+            git_dep = GitDependency(
+                name=gem.name,
+                # https://github.com/containerbuildsystem/gemlock-parser/blob/master/gemlock_parser/gemfile_lock.py#L185
+                version=gem.spec_version,
+                url=gem.remote,
+                revision=gem.revision,
+                branch=gem.branch,
+            )
+            _validate_git_dependency(git_dep)
```

Comment:
Good question, again, most likely consistency with other pkg managers, but my take on that is that since we're adding a new one, we might as well define an abstract interface for these dep types and let inheritance take care of this like you say.

Reply:
Previously, I used Pydantic base models instead of  `@dataclass`. Would pydantic validators solve this, or am I missing something?

Reply:
Yes, `@field_validator` should do the trick.

Reply:
Agreed, this is a good example of post-init validation (the easier use case vs compared to the RPM input JSON model) so field validator is enough.